regex-syntax/src/parser.rs

```diff
@@ -526,65 +526,132 @@ impl Parser {
+                    class = class.intersection(&class2);
+                }
+                _ => unreachable!()
+            }
+        }
+
+        // negate after combining all sets (`^` has lower precedence than `&&`)
+        if negated {
+            class = class.negate();
+        }
+        if class.is_empty() {
+            // e.g., [^\d\D]
+            return Err(self.err(ErrorKind::EmptyClass));
+        }
+        Ok(class)
+    }
+
+    // Parses a set in a character class. A set is the union of multiple
+    // consecutive single characters, ranges or nested sets.
+    //
+    // Terminates either when encountering a closing `]` or a `&&`.
+    //
+    // The returned `CharClass` is canonical.
+    //
+    // e.g., `[a-cd&&x-z]`
+    //
+    // Start: `a` (with first_set == true)
+    // End:   `&` (the first one)
+    //
+    // or
+    //
+    // Start: `x` (with first_set == false)
+    // End:   `]`
+    fn parse_class_set(&mut self, first_set: bool) -> Result<CharClass> {
         let mut class = CharClass::empty();
-        while self.bump_if('-') {
-            class.ranges.push(ClassRange::one('-'));
+        // `-` at the start of the first set in a class is allowed.
+        // If it occurs after a `&&`, we need to check for the `--` operator.
+        if first_set {
+            while self.bump_if('-') {
+                class.ranges.push(ClassRange::one('-'));
+            }
         }
         loop {
             if self.eof() {
                 // e.g., [a
                 return Err(self.err(ErrorKind::UnexpectedClassEof));
             }
             match self.cur() {
-                // If no ranges have been added, then `]` is the first
-                // character (sans, perhaps, the `^` symbol), so it should
+                // If we're at the start of the class and `]` is the first
+                // character (sans, perhaps, the `^` symbol), it should
                 // be interpreted as a `]` instead of a closing class bracket.
-                ']' if class.len() > 0 => { self.bump(); break }
+                // If we're after `&&` or there were other characters already,
+                // it should be interpreted as the end of the class.
+                ']' if !(first_set && class.len() == 0) => { break }
+                // `&&` means intersection with the next set, so stop parsing.
+                '&' if self.peek_is("&&") => { break }
                 '[' => match self.maybe_parse_ascii() {
                     Some(class2) => class.ranges.extend(class2),
                     None => {
-                        return Err(self.err(
-                            ErrorKind::UnsupportedClassChar('[')));
-                    }
-                },
-                '\\' => match try!(self.parse_escape()) {
-                    Build::Expr(Expr::Class(class2)) => {
+                        // Nested set, e.g. `[c-d]` in `[a-b[c-d]]`
+                        let class2 = try!(self.parse_class_as_chars());
```

Comment:
Very good point. I remember wondering about that problem while implementing it, but then forgot later.

I'll look into using `Build`, yeah. Maybe it could also be a separate stack. Have to wrap my head around it first though :).